spec/memory-safe-d.dd

```diff
@@ -24,19 +24,53 @@ $(HEADERNAV_TOC)
 
 $(H2 $(LNAME2 usage, Usage))
 
-        $(P Memory safety can be enabled on a per-function basis using
-        the $(DDSUBLINK spec/function, safe-functions, $(D @safe) attribute).
-        This can be inferred when the compiler has the function body
-        available. The $(DDSUBLINK spec/function, trusted-functions, `@trusted` attribute) can be used when a function has a safe
-        interface, but uses unsafe code internally. These functions can
-        be called from $(D @safe) code.
+        $(P There are three categories of functions from the perspective of memory safety:)
+        $(UL
+                $(LI $(DDSUBLINK spec/function, safe-functions, $(D @safe)) functions)
+                $(LI $(DDSUBLINK spec/function, trusted-functions, $(D `@trusted`)) functions)
+                $(LI $(DDSUBLINK spec/function, system-functions, $(D `@system`)) functions)
         )
 
+        $(P $(D `@system`) functions may perform any operation legal from the perspective of the language including inherently
+        memory unsafe operations like returning pointers to expired stackframes. These functions may not be called directly from
+        $(D `@safe`) functions.)
+
+        $(P $(D `@trusted`) functions have all the capabilities of $(D `@system`) functions but may be called from
+        $(D @safe) functions. For this reason they should be very limited in the scope of their use. Typical uses of
+        $(D `@trusted`) functions include wrapping system calls that take buffer pointer and length arguments separately so that
+        $(D @safe) functions may call them with arrays.)
+
+        $(P $(D @safe) functions have a number of restrictions on what they may do and are intended to disallow operations that
+        may cause memory corruption. See $(DDSUBLINK spec/function, safe-functions, $(D @safe) functions).)
+
+        $(P These attributes may be inferred when the compiler has the function body available, such as with templates.)
+
         $(P Array bounds checks are necessary to enforce memory safety, so
         these are enabled (by default) for $(D @safe) code even in $(B
         -release) mode.
         )
 
+    $(H3 $(LNAME2 scope-return-params, Scope & Return Parameters))
+
+         $(P The function parameter attributes $(D return) and $(D scope) are used to track what happens to low-level pointers 
+         passed to functions. Such pointers include: pointers, $(D this), classes, `ref` parameters, delegate/lazy parameters, 
+         and aggregates containing a pointer.)
```

Comment:
Arrays? Or do you count those as aggregates?

Reply:
Suggest we use `raw pointers, this, variables of reference type, ref/lazy parameters, ref return values and aggregates containing such pointers`.